# UMP16: Zuordnungsproblem Aufgaben / Arbeiter

## Lösung ILP in Matrix Notation
Nachfolgendes Skript zeigt die Lösung des Modells als ILP, d.h. als ganzzahlig lineares Programm, ausgehend von einer Formulierung in **Matrix Notation**. 

Dieses Beispiel dient als Abschreckung. Es zeigt, wie schnell (manuell erzeugte) Matrix-Notation unübersichtlich wird. Das möchte man nicht von hand programmieren müssen. Auch ist es sehr mühsam und bedeutet viel Fleiss, die erstellten ILP Modelle zuerst in die benötigte Form zu bringen, damit sie von den Matrix-Solvern gelöst werden können.
Aus diesem Grund werden grössere Modelle i.A. auf Basis von algebraischer Notation formuliert und gelöst, wie wir dies im übernächsten Skript-Teil demonstrieren. Jetzt aber zuerst der Matrix-Ansatz:


In [1]:
from UMP16_ILP_MATRIX_NOTATION import *

main()

Optimization terminated successfully. (HiGHS Status 7: Optimal)
Objective value:	650.0
Results:			[ 0. -0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  0. -0.  0.
  0.  1.  1. -0. -0.  0.  0.]


Es konnte die optimale Lösung gefunden werden. Der Ergebnisvektor muss noch interpretiert werden. Einzelne Werte enthalten -0, sind aber als 0 zu interpretieren. Die Ursache dafür liegt in numerischen Eigenheiten des Solvers. Scipy verwendet auch bei Integralen Constraints Gleitkomma-Zahlen, die in diesem Falle sehr nahe bei 0 liegen, aber dennoch nicht exakt gleich 0 sind. Für unsere Interpretation sind die Werte jedoch gleichbedeutend mit 0.

Lösung: Arbeiter 1 sollte Aufgabe 4, Arbeiter 2 soll Aufgabe 3, Arbeiter 3 soll Aufgabe 2, Arbeiter 4 soll Aufgabe 5 und Arbeiter 1 soll Aufgabe 1 lösen. Dadurch entstehen minimale Kosten von insgesamt 650 CHF. 

## Lösung ILP in algebraischer Notation
Nachfolgendes Skript zeigt die Lösung des Modells als ILP, d.h. als ganzzahlig lineares Programm, ausgehend von einer Formulierung in **algebraischer Notation**. Beachten Sie unbedingt im Quellcode die viel einfachere Formulierung des Modells, sowie die intuitivere Angabe der zur Verfügung stehenden Parameter. 

In [3]:
from UMP16_ILP_ALGEBRAIC_NOTATION import *

main()

Solution:
Objective value = 650.0
X_0_0 = 0
X_0_1 = 0
X_0_2 = 0
X_0_3 = 1
X_0_4 = 0
X_1_0 = 0
X_1_1 = 0
X_1_2 = 1
X_1_3 = 0
X_1_4 = 0
X_2_0 = 0
X_2_1 = 1
X_2_2 = 0
X_2_3 = 0
X_2_4 = 0
X_3_0 = 0
X_3_1 = 0
X_3_2 = 0
X_3_3 = 0
X_3_4 = 1
X_4_0 = 1
X_4_1 = 0
X_4_2 = 0
X_4_3 = 0
X_4_4 = 0


Die Lösung bleibt sich natürlich gleich, egal welche der beiden Formulierungen (Matrix- oder algebraische Notation) wir wählen. Zu bedenken ist bei der Interpretation der Lösung, dass die Indizes in Python bei 0 beginnen, nicht bei 1.

Lösung: Arbeiter 1 sollte Aufgabe 4, Arbeiter 2 soll Aufgabe 3, Arbeiter 3 soll Aufgabe 2, Arbeiter 4 soll Aufgabe 5 und Arbeiter 1 soll Aufgabe 1 lösen. Dadurch entstehen minimale Kosten von insgesamt 650 CHF. 